mirrord/layer-lib/src/process/windows/execution/mod.rs

```diff
@@ -0,0 +1,458 @@
+        },
+        synchapi::WaitForSingleObject,
+        winbase::{CREATE_SUSPENDED, CREATE_UNICODE_ENVIRONMENT, INFINITE, WAIT_OBJECT_0},
+        wincon::FreeConsole,
+        winnt::PHANDLE,
+    },
+};
+
+use super::sync::{LayerInitEvent, MIRRORD_LAYER_INIT_EVENT_NAME};
+use crate::{
+    error::{LayerError, LayerResult, windows::WindowsError},
+    proxy_connection::PROXY_CONNECTION,
+    setup::layer_setup,
+    socket::sockets::SHARED_SOCKETS_ENV_VAR,
+};
+
+pub mod debug;
+
+pub use debug::{
+    format_debugger_config, get_current_process_name, is_debugger_wait_enabled,
+    should_wait_for_debugger,
+};
+
+pub const MIRRORD_AGENT_ADDR_ENV: &str = "MIRRORD_AGENT_ADDR";
+pub const MIRRORD_LAYER_ID_ENV: &str = "MIRRORD_LAYER_ID";
+pub const MIRRORD_LAYER_FILE_ENV: &str = "MIRRORD_LAYER_FILE";
+
+// Windows-specific child process inheritance environment variables
+pub const MIRRORD_LAYER_CHILD_PROCESS_PARENT_PID: &str = "MIRRORD_LAYER_CHILD_PROCESS_PARENT_PID";
+pub const MIRRORD_LAYER_CHILD_PROCESS_LAYER_ID: &str = "MIRRORD_LAYER_CHILD_PROCESS_LAYER_ID";
+pub const MIRRORD_LAYER_CHILD_PROCESS_PROXY_ADDR: &str = "MIRRORD_LAYER_CHILD_PROCESS_PROXY_ADDR";
+
+const LAYER_INIT_TIMEOUT_MS: u32 = 30_000;
+
+/// Environment variables that should be explicitly forwarded from parent to child process
+const FORWARDED_ENV_VARS: &[&str] = &[
+    MIRRORD_AGENT_ADDR_ENV,
+    MIRRORD_LAYER_ID_ENV,
+    MIRRORD_LAYER_FILE_ENV,
+    "RUST_LOG",
+    "RUST_BACKTRACE",
+];
+
+/// Function signature for CreateProcessInternalW Windows API call.
+pub type CreateProcessInternalWType = unsafe extern "system" fn(
+    user_token: HANDLE,
+    application_name: LPCWSTR,
+    command_line: LPWSTR,
+    process_attributes: LPSECURITY_ATTRIBUTES,
+    thread_attributes: LPSECURITY_ATTRIBUTES,
+    inherit_handles: BOOL,
+    creation_flags: DWORD,
+    environment: LPVOID,
+    current_directory: LPCWSTR,
+    startup_info: LPSTARTUPINFOW,
+    process_information: LPPROCESS_INFORMATION,
+    restricted_user_token: PHANDLE,
+) -> BOOL;
+
+pub struct LayerManagedProcess {
+    process_info: PROCESS_INFORMATION,
+    released: bool,
+    terminate_on_drop: bool,
+}
+
+impl LayerManagedProcess {
+    /// Build Windows environment block from HashMap (UTF-16 format)
+    fn build_windows_env_block(environment: &HashMap<String, String>) -> Vec<u16> {
+        let mut windows_environment: Vec<u16> = Vec::new();
+        for (key, value) in environment {
+            let entry = format!("{}={}", key, value);
+            let entry_wide = string_to_u16_buffer(&entry);
+            windows_environment.extend(entry_wide);
+        }
+        // add environment block null terminator
+        windows_environment.push(0);
+        windows_environment
+    }
+
+    /// Add mirrord-specific environment variables to caller's environment
+    fn add_mirrord_env_vars(env_vars: &mut HashMap<String, String>, parent_event: &LayerInitEvent) {
```

Comment:
what happens when a child process calls CreateProcessW with the environment variable being null, causing environment to be inherited?

do we have any handling so that `add_mirrord_env_vars` doesn't cause duplicated environment keys? can we inherit child process env vars again, then also insert them, in a child tree like the following?

- process1.exe
  - process2.exe (child)
    - process3.exe (child of process2.exe, which is a child of process1.exe)